File: source/com/barrybecker4/puzzle/redpuzzle/model/PieceList.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.puzzle.redpuzzle.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;

/**
 * The pieces that are in the red puzzle.
 * model in the model-view-controller pattern.
 * This class is only mostly immutable.
 * There is a cost for immutability, and I did not do it for methods that are not needed for concurrency safety.
 *
 * @author Barry Becker
 */
public class PieceList {

    /** the real game has 9 pieces, but I might experiment with 4 or 16 for testing. */
    public static final int DEFAULT_NUM_PIECES = 9;

    /** use the same seed for repeatable results. */
    private static final int SEED = 5;

    /** There are either 4 or 9 or 16 pieces. */
    private int numPieces = DEFAULT_NUM_PIECES;
    private int edgeLen;

    /**
     * 0 solves in 25,797 tries.
     * 1 solves in 15,444
     * 2 solves in 22,293
     * 3 solves in 21,588
     * 4 solves in  4,005
     * 5 solves in  5,319
     */
    private static Random RANDOM = new Random(SEED);

    private List<Piece> pieces_;


    /**
     * a list of 9 puzzle pieces.
     */
    public PieceList() {
        this(DEFAULT_NUM_PIECES);
    }

    /**
     * a list of puzzle pieces (initially empty).
     */
    private PieceList(int numPieces) {

        if (numPieces == 0)
            throw new IllegalArgumentException("Num pieces should not be 0");
        assert(numPieces==4 || numPieces == 9 || numPieces == 16);
        this.numPieces = numPieces;
        this.edgeLen = (int)Math.sqrt(numPieces);
        pieces_ = new LinkedList<>();
    }

    /**
     * a list of puzzle pieces with pieces to use specified.
     * @param pieces array of pieces to use.
     */
    public PieceList(Piece[] pieces) {
        this(pieces.length);
        addPieces(pieces);
    }

    /**
     * Copy constructor
     */
    public PieceList(PieceList pieces) {
        this(pieces.pieces_, pieces.getTotalNum());
    }

    /**
     * private copy constructor
     */
    private PieceList(List<Piece> pieces, int numTotal) {
        this(numTotal);
        for (Piece p : pieces) {
            pieces_.add(new Piece(p));
        }
    }

    private void addPieces(Piece[] pieces) {
        Collections.addAll(pieces_, pieces);
    }

    /**
     * @param i the index of the piece to get.
     * @return the i'th piece.
     */
    public Piece get(int i)  {
        assert i < numPieces :
                "there are only " + numPieces + " pieces, but you tried to get the " + (i+1) + "th";
        return pieces_.get(i);
    }

    /**
     * @return the last piece added.
     */
    public Piece getLast()  {

        return pieces_.get(pieces_.size() - 1);
    }

    /**
     * @return The total number of pieces in the puzzle.
     */
    public int getTotalNum() {
        return numPieces;
    }

    /**
     * @return the number of pieces in the list.
     */
    public int size() {
        return pieces_.size();
    }

    /**
     * Does this need to be made immutable?
     * Swap 2 pieces.
     */
    public void doSwap(int p1Pos, int p2Pos) {
       assert p1Pos <= numPieces&& p2Pos < numPieces :
                "The position indices must be less than " + numPieces + ".  You had " + p1Pos + ",  " + p2Pos;
        Piece piece2 = get(p2Pos);
        Piece piece1 = pieces_.remove(p1Pos);
        pieces_.add(p1Pos, piece2);
        pieces_.remove(piece2);
        pieces_.add(p2Pos, piece1);
    }

    /**
     * @param p piece to add to the end of the list.
     */
    public PieceList add(Piece p) {
        return add(pieces_.size(), p);
    }

    /**
      * @param p piece to add to the end of the list.
      */
     public PieceList add(int i, Piece p) {
         PieceList newPieceList = new PieceList(this);
         newPieceList.pieces_.add(i, p);
         assert newPieceList.pieces_.size() <= numPieces :
                "there can only be at most " + numPieces + " pieces.";
         return newPieceList;
    }

    /**
     * @param p the piece to remove.
     * @return true if the list contained this element.
     */
    public PieceList remove(Piece p) {
        PieceList newPieceList = new PieceList(this);
        boolean removed = newPieceList.pieces_.remove(p);
        assert removed: " could not remove " + p + " from " + newPieceList.pieces_;
        return newPieceList;
    }

    /**
     * @return piece list after removing the last element.
     */
    public PieceList removeLast() {
        return remove(pieces_.size() - 1);
    }

    /**
     * @return the list with the ith element removed.
     */
    public PieceList remove(int index) {
         PieceList newPieceList = new PieceList(this);
         newPieceList.pieces_.remove(index);
         return newPieceList;
    }

    /**
     * ? Does this need to be made immutable?
     */
    public void rotate(int k, int numRotations) {
        Piece p = pieces_.get(k);
        Piece rp = p.rotate(numRotations);
        pieces_.set(k, rp);
    }

    /**
     *@return a new shuffled PieceList object based on the old.
     */
    public PieceList shuffle() {
        // randomly rotate all the pieces
        List<Piece> plist = new ArrayList<>(pieces_.size());
        for (Piece p : pieces_) {
            Piece newp = p.rotate(RANDOM.nextInt(4));
            plist.add(newp);
        }
        Collections.shuffle(plist, RANDOM);
        return new PieceList(plist, numPieces);
    }

    /**
     * Try the piece.
     * @param piece the piece to try to fit into our current solution.
     * @return true if it fits.
     */
    public boolean fits(Piece piece) {

        // it needs to match the piece to the left and above (if present)
        boolean fits = true;
        int numSolved = size();
        int row = numSolved / edgeLen;
        int col = numSolved % edgeLen;
        if ( col > 0 ) {
            // if other than a left edge piece, then we need to match to the left side nub.
            Piece leftPiece = getLast();
            if (!leftPiece.getRightNub().fitsWith(piece.getLeftNub()))
                fits = false;
        }
        if ( row > 0 ) {
            // then we need to match with the top one
            Piece topPiece = get( numSolved - edgeLen );
            if (!topPiece.getBottomNub().fitsWith(piece.getTopNub()) )
                fits = false;
        }

        return fits;
    }

    /**
     * @return the number of matches for the nubs on this piece
     */
    public int getNumFits(int i) {

        assert(i < numPieces);
        // it needs to match the piece to the left and above (if present)
        int numFits = 0;
        int dim = edgeLen;
        int row = i / dim;
        int col = i % dim;
        Piece piece = get(i);

        if ( col > 0 ) {
            // if other than a left edge piece, then we need to match to the left side nub.
            Piece leftPiece = get( i - 1 );
            if (leftPiece.getRightNub().fitsWith(piece.getLeftNub()))
                numFits++;
        }
        if ( row > 0 ) {
            // then we need to match with the top one
            Piece topPiece = get( i - dim );
            if (topPiece.getBottomNub().fitsWith(piece.getTopNub()))
                numFits++;
        }

        if ( col < (dim-1) ) {
            // if other than a right edge piece, then we need to match to the right side nub.
            Piece rightPiece = get( i + 1 );
            if (rightPiece.getLeftNub().fitsWith(piece.getRightNub()))
                numFits++;
        }
        if ( row < (dim-1) ) {
            // then we need to match with the bottom one
            Piece bottomPiece = get( i + dim );
            if (bottomPiece.getTopNub().fitsWith(piece.getBottomNub()))
                numFits++;
        }

        return numFits;
    }

    /**
     *@return as unmodifiable list so there can be no malicious modification of our immutable state.
     */
    public List<Piece> getPieces() {
        return Collections.unmodifiableList(pieces_);
    }

    /**
     *@return true if we contain p
     */
    public boolean contains(Piece p) {
        return pieces_.contains(p);
    }

    @Override
    public String toString() {
        StringBuilder buf = new StringBuilder("PieceList: ("+size()+" pieces)\n");
        for (Piece p : pieces_) {
            buf.append(p.toString()).append('\n');
        }
        return buf.toString();
    }
}

File: scala-source/com/barrybecker4/puzzle/redpuzzle/model/PieceList.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.redpuzzle.model

import com.barrybecker4.math.MathUtil

import scala.util.Random


object PieceList {

  /** The standard game has 9 pieces; 4 pieces are supported for smaller puzzles. */
  val DEFAULT_NUM_PIECES = 9

  /** Row and column (0-based) for a linear index on a square grid of `pieceCount` pieces. */
  def rowColForIndex(index: Int, pieceCount: Int): (Int, Int) = {
    val edge = Math.sqrt(pieceCount).toInt
    (index / edge, index % edge)
  }
}

/**
  * A list of fit puzzle pieces (initially empty) for a numTotal sized puzzle.
  * @param pieces the currently fit pieces. May be be less than numTotal of them, but not more,
  * @param numTotal the total number of pieces in the puzzle.
  * @author Barry Becker
  */
case class PieceList(pieces: List[OrientedPiece], numTotal: Int) {

  require(numTotal == 4 || numTotal == 9, s"$numTotal is not a valid number of pieces")
  val edgeLength: Int = Math.sqrt(numTotal).toInt


  def this(pieces: Array[Piece]) = {
    this(pieces.map(p => OrientedPiece(p)).toList, pieces.length)
  }

  def this(numTotal: Int = PieceList.DEFAULT_NUM_PIECES) = {
    this(List[OrientedPiece](), numTotal)
  }

  /** @param i the index of the piece to get.
    * @return the i'th piece.
    */
  def get(i: Int): OrientedPiece = {
    assert(i < numTotal, "there are only " + numTotal + " pieces, but you tried to get the " + (i + 1) + "th")
    pieces(i)
  }

  /** @return the last piece added. */
  def getLast: OrientedPiece = pieces.last

  /** @return the number of pieces in the list.*/
  def size: Int = pieces.size

  def isEmpty: Boolean = pieces.isEmpty

  /** @return new piece list with 2 indicated pieces swapped. */
  def doSwap(p1Pos: Int, p2Pos: Int): PieceList = {
    require(
      p1Pos >= 0 && p1Pos < numTotal && p2Pos >= 0 && p2Pos < numTotal,
      s"Swap indices must be in [0, $numTotal); got $p1Pos, $p2Pos"
    )
    PieceList(pieces.updated(p2Pos, pieces(p1Pos)).updated(p1Pos, pieces(p2Pos)), numTotal)
  }

  /** @param p piece to add to the end of the list. */
  def add(p: OrientedPiece): PieceList = add(pieces.size, p)

  /** @param position the (1 based) position to add the piece.
    * @param p piece to add at the specified position in the list.
    */
  def add(position: Int, p: OrientedPiece): PieceList = {
    require(size < numTotal, "Cannot add another piece when already at max of " + numTotal)
    val (front, back) = pieces.splitAt(position)
    PieceList(front ++ List(p) ++ back, numTotal)
  }

  def remove(p: Piece): PieceList = PieceList(pieces.filter(x => x.piece != p), numTotal)

  def removeLast(): PieceList = remove(pieces.size - 1)

  /** @return the list with the ith element removed. */
  def remove(index: Int): PieceList = PieceList(pieces.take(index) ++ pieces.drop(index + 1), numTotal)

  def rotate(k: Int, numRotations: Int): PieceList = {
    val rp = pieces(k).rotate(numRotations)
    PieceList(pieces.updated(k, rp), numTotal)
  }

  /** @return a new shuffled PieceList object based on the old. All pieces moved and rotated */
  def shuffle(rnd: Random = MathUtil.RANDOM): PieceList = {
    val pieceList = for (p <- pieces) yield p.rotate(rnd.nextInt(4))
    PieceList(rnd.shuffle(pieceList), numTotal)
  }

  /** @param piece the piece to try to fit into our current solution (by checking the next available position).
    * @return true if the piece fits.
    */
  def fits(piece: OrientedPiece): Boolean = {
    val numSolved = size
    val col = numSolved % edgeLength
    val row = numSolved / edgeLength
    val leftOk = col == 0 || nubsFitHorizontally(getLast, piece)
    val topOk = row == 0 || nubsFitVertically(get(numSolved - edgeLength), piece)
    leftOk && topOk
  }

  /** @return the number of matches for the nubs on this piece (at most 4) */
  def getNumFits(pos: Int): Int = {
    assert(pos < numTotal)
    val row = pos / edgeLength
    val col = pos % edgeLength
    val piece = get(pos)
    var count = 0
    if col > 0 && nubsFitHorizontally(get(pos - 1), piece) then count += 1
    if row > 0 && nubsFitVertically(get(pos - edgeLength), piece) then count += 1
    if col < edgeLength - 1 && nubsFitHorizontally(piece, get(pos + 1)) then count += 1
    if row < edgeLength - 1 && nubsFitVertically(piece, get(pos + edgeLength)) then count += 1
    count
  }

  /** @return true if we contain p */
  def contains(p: Piece): Boolean = pieces.exists(_.piece == p)

  override def toString: String = {
    val buf = new StringBuilder("PieceList: (" + size + " pieces)\n")
    for (p <- pieces) buf.append(p.toString).append('\n')
    buf.toString
  }

  private def nubsFitHorizontally(left: OrientedPiece, right: OrientedPiece): Boolean =
    left.getRightNub.fitsWith(right.getLeftNub)

  private def nubsFitVertically(above: OrientedPiece, below: OrientedPiece): Boolean =
    above.getBottomNub.fitsWith(below.getTopNub)
}
